src/lib.rs

```diff
@@ -130,61 +123,49 @@ where
     }
 }
 
-impl<R> Read for TimeoutReader<R>
-where
-    R: AsyncRead,
-{
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        let r = self.reader.read(buf);
-        match r {
-            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => self.state.check()?,
-            _ => self.state.reset(),
-        }
-        r
-    }
-}
-
 impl<R> AsyncRead for TimeoutReader<R>
 where
-    R: AsyncRead,
+    R: AsyncRead + Unpin,
 {
-    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
-        self.reader.prepare_uninitialized_buffer(buf)
-    }
-
-    fn read_buf<B: BufMut>(&mut self, buf: &mut B) -> Poll<usize, io::Error> {
-        let r = self.reader.read_buf(buf);
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let s = self.get_mut();
+        let r = Pin::new(&mut s.reader).poll_read(cx, buf);
         match r {
-            Ok(Async::NotReady) => self.state.check()?,
-            _ => self.state.reset(),
+            Poll::Pending => s.state.poll_check(cx)?,
+            _ => s.state.reset(),
         }
         r
     }
```

Comment:
What about poll_read_buf/poll_write_buf/prepare_uninitialized_buffer?

Reply:
I can implement them if you'd like. Since the default implementations in the respective traits are implemented in terms of these primitive trait functions, what's the benefit?

Reply:
If the inner stream overrides those methods, we want the behavior to be maintained.

Reply:
Rust is still a bit new to me, so please excuse me for still not understanding. How can the AsyncWrite implementation by the contained stream override the default methods in our implementation?  I wrote the following test which behaves like I would expect:

```
    impl AsyncRead for DelayStream {
        unsafe fn prepare_uninitialized_buffer(&self, _buf: &mut [u8]) -> bool {
            panic!("Die!");
        }
        // ...
    }

    #[tokio::test]
    async fn read_ok() {
        let reader = DelayStream(delay(Instant::now() + Duration::from_millis(100)));
        let mut buf: [u8; 1] = [0; 1];
        let mut reader = TimeoutReader::new(reader);

        // this passes just fine, but panics if I put it before wrapping it
        assert!(unsafe {reader.prepare_uninitialized_buffer(&mut buf)});

        reader.set_timeout(Some(Duration::from_millis(500)));
        read_one(reader).await.unwrap();
    }
````
I could see a consumer overriding these functions in our trait, but us specifying them doesn't stop that behavior, correct?